domain/application/state/application.go

```diff
@@ -821,45 +802,18 @@ func (st *State) UpdateCAASUnit(ctx context.Context, unitName coreunit.Name, par
 				return errors.Errorf("updating cloud container for unit %q: %w", unitName, err)
 			}
 		}
-		if params.AgentStatus != nil {
-			if err := st.setUnitAgentStatus(ctx, tx, toUpdate.UnitUUID, application.UnitAgentStatusInfo{
-				StatusID: application.MarshallUnitAgentStatus(params.AgentStatus.Status),
-				StatusInfo: application.StatusInfo{
-					Message: params.AgentStatus.Message,
-					Data: transform.Map(
-						params.AgentStatus.Data, func(k string, v any) (string, string) { return k, fmt.Sprint(v) }),
-					Since: since(params.AgentStatus.Since),
-				},
-			}); err != nil {
-				return errors.Errorf("saving unit %q agent status: %w", unitName, err)
-			}
+
+		if err := st.setUnitAgentStatus(ctx, tx, toUpdate.UnitUUID, params.AgentStatus); err != nil {
+			return errors.Errorf("saving unit %q agent status: %w", unitName, err)
 		}
-		if params.WorkloadStatus != nil {
-			if err := st.setUnitWorkloadStatus(ctx, tx, toUpdate.UnitUUID, application.UnitWorkloadStatusInfo{
-				StatusID: application.MarshallUnitWorkloadStatus(params.WorkloadStatus.Status),
-				StatusInfo: application.StatusInfo{
-					Message: params.WorkloadStatus.Message,
-					Data: transform.Map(
-						params.WorkloadStatus.Data, func(k string, v any) (string, string) { return k, fmt.Sprint(v) }),
-					Since: since(params.WorkloadStatus.Since),
-				},
-			}); err != nil {
-				return errors.Errorf("saving unit %q workload status: %w", unitName, err)
-			}
+
+		if err := st.setUnitWorkloadStatus(ctx, tx, toUpdate.UnitUUID, params.WorkloadStatus); err != nil {
+			return errors.Errorf("saving unit %q workload status: %w", unitName, err)
 		}
-		if params.CloudContainerStatus != nil {
-			if err := st.setCloudContainerStatus(ctx, tx, toUpdate.UnitUUID, application.CloudContainerStatusStatusInfo{
-				StatusID: application.MarshallCloudContainerStatus(params.CloudContainerStatus.Status),
-				StatusInfo: application.StatusInfo{
-					Message: params.CloudContainerStatus.Message,
-					Data: transform.Map(
-						params.CloudContainerStatus.Data, func(k string, v any) (string, string) { return k, fmt.Sprint(v) }),
-					Since: since(params.CloudContainerStatus.Since),
-				},
-			}); err != nil {
-				return errors.Errorf("saving unit %q cloud container status: %w", unitName, err)
-			}
+		if err := st.setCloudContainerStatus(ctx, tx, toUpdate.UnitUUID, params.CloudContainerStatus); err != nil {
+			return errors.Errorf("saving unit %q cloud container status: %w", unitName, err)
 		}
```

Comment:
Is this ok here, or should it be moved into the `if cloudContainer != nil {` block?

I see these method start wit ha catch guard so I'm sure technically it is, but I mean conceptually

Reply:
This method was just a massive eyesore, we should revisit it.